drivers/can/can_nxp_s32_canxl.c

```diff
@@ -0,0 +1,977 @@
+/*
+ * Copyright 2022 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/sys/atomic.h>
+#include <zephyr/drivers/can.h>
+#include <zephyr/drivers/can/transceiver.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/logging/log.h>
+
+#include <CanEXCEL_Ip_HwAccess.h>
+#include <CanEXCEL_Ip_Irq.h>
+#include "can_nxp_s32_canxl.h"
+
+#define DT_DRV_COMPAT nxp_s32_canxl
+
+#define CAN_NXP_S32_TIMEOUT_MS  1
+#define CAN_NXP_S32_MAX_BITRATE	16000000
+
+LOG_MODULE_REGISTER(nxp_s32_canxl, CONFIG_CAN_LOG_LEVEL);
+
+#define SP_AND_TIMING_NOT_SET(inst)				\
+	(!DT_INST_NODE_HAS_PROP(inst, sample_point) &&		\
+	!(DT_INST_NODE_HAS_PROP(inst, prop_seg) &&		\
+	DT_INST_NODE_HAS_PROP(inst, phase_seg1) &&		\
+	DT_INST_NODE_HAS_PROP(inst, phase_seg2))) ||
+
+#if DT_INST_FOREACH_STATUS_OKAY(SP_AND_TIMING_NOT_SET) 0
+#error You must either set a sampling-point or timings (phase-seg* and prop-seg)
+#endif
+
+#ifdef CONFIG_CAN_FD_MODE
+
+#define SP_AND_TIMING_DATA_NOT_SET(inst)			\
+	(!DT_INST_NODE_HAS_PROP(inst, sample_point_data) &&	\
+	!(DT_INST_NODE_HAS_PROP(inst, prop_seg_data) &&		\
+	DT_INST_NODE_HAS_PROP(inst, phase_seg1_data) &&		\
+	DT_INST_NODE_HAS_PROP(inst, phase_seg2_data))) ||
+
+#if DT_INST_FOREACH_STATUS_OKAY(SP_AND_TIMING_DATA_NOT_SET) 0
+#error You must either set a sampling-point-data or timings (phase-seg-data* and prop-seg-data)
+#endif
+#endif
+
+
+static int can_nxp_s32_get_capabilities(const struct device *dev, can_mode_t *cap)
+{
+	ARG_UNUSED(dev);
+
+	*cap = CAN_MODE_NORMAL | CAN_MODE_LOOPBACK | CAN_MODE_LISTENONLY | CAN_MODE_FD;
+
+	return 0;
+}
+
+static int can_nxp_s32_start(const struct device *dev)
+{
+	const struct can_nxp_s32_config *config = dev->config;
+	struct can_nxp_s32_data *data = dev->data;
+	int err;
+
+	if (data->started) {
+		return -EALREADY;
+	}
+
+	Canexcel_Ip_EnterFreezeMode(config->instance);
+
+	if (config->phy != NULL) {
+		err = can_transceiver_enable(config->phy);
+		if (err != 0) {
+			LOG_ERR("failed to enable CAN transceiver (err %d)", err);
+			return err;
+		}
+	}
+
+	Canexcel_Ip_ExitFreezeMode(config->instance);
```

Comment:
Is it necessary to enter freeze mode in order to enable the PHY? If so, why?

This function is supposed to start the CAN controller so that it begins sending ACKs for received frames. I am unable to tell if this is done here?

Reply:
yes, I removed